File: android/app/src/main/java/com/clipsync/android/protocol/PairAuthProof.kt
package com.clipsync.android.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Pair-secret challenge-response proof from docs/protocol-v1.md section 3.
 * Port of windows/ClipSync.Core/Security/PairAuthProof.cs. Shared vectors live in
 * protocol/v1/fixtures/auth/vectors.json.
 *
 * UUID_BYTES is RFC 4122 big-endian from the 32 hex chars of the canonical UUID,
 * not Java UUID most/least-significant-bit layout.
 */
object PairAuthProof {
    const val ALGORITHM = "hmac-sha256"
    const val SECRET_LENGTH = 32
    const val NONCE_LENGTH = 32
    const val PROOF_LENGTH = 32

    private val PREFIX_V1 = "ClipSync/v1/auth\n".toByteArray(Charsets.UTF_8)
    private val PREFIX_V2 = "ClipSync/v2/auth\n".toByteArray(Charsets.UTF_8)

    fun compute(
        pairSecret: ByteArray,
        challengeRequestId: String,
        nonce: ByteArray,
        challengerDeviceId: String,
        responderDeviceId: String,
        trustEpoch: Long,
        protocolVersion: Int = 1,
    ): ByteArray {
        require(pairSecret.size == SECRET_LENGTH) { "The pair secret must be exactly 32 bytes." }
        require(nonce.size == NONCE_LENGTH) { "The challenge nonce must be exactly 32 bytes." }
        require(protocolVersion == 1 || protocolVersion == 2) { "Protocol version must be 1 or 2." }

        val prefix = if (protocolVersion == 2) PREFIX_V2 else PREFIX_V1
        val requestIdBytes = canonicalUuid(challengeRequestId).toByteArray(Charsets.UTF_8)
        val extra = if (protocolVersion == 2) 1 + 8 else 0
        val message = ByteArray(
            prefix.size + requestIdBytes.size + 1 + NONCE_LENGTH + 16 + 16 + 8 + extra,
        )
        var offset = 0
        prefix.copyInto(message, offset)
        offset += prefix.size
        requestIdBytes.copyInto(message, offset)
        offset += requestIdBytes.size
        message[offset] = 0x00
        offset += 1
        nonce.copyInto(message, offset)
        offset += NONCE_LENGTH
        uuidBytes(challengerDeviceId).copyInto(message, offset)
        offset += 16
        uuidBytes(responderDeviceId).copyInto(message, offset)
        offset += 16
        ByteBuffer.wrap(message, offset, 8).order(ByteOrder.BIG_ENDIAN).putLong(trustEpoch)
        offset += 8
        if (protocolVersion == 2) {
            message[offset] = 0x00
            offset += 1
            ByteBuffer.wrap(message, offset, 8).order(ByteOrder.BIG_ENDIAN).putLong(protocolVersion.toLong())
        }

        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(pairSecret, "HmacSHA256"))
        return mac.doFinal(message)
    }

    fun verify(
        pairSecret: ByteArray,
        challengeRequestId: String,
        nonce: ByteArray,
        challengerDeviceId: String,
        responderDeviceId: String,
        trustEpoch: Long,
        proof: ByteArray,
        protocolVersion: Int = 1,
    ): Boolean {
        if (proof.size != PROOF_LENGTH) {
            return false
        }
        val expected = compute(
            pairSecret,
            challengeRequestId,
            nonce,
            challengerDeviceId,
            responderDeviceId,
            trustEpoch,
            protocolVersion,
        )
        return MessageDigest.isEqual(expected, proof)
    }

    fun encodeBase64Url(bytes: ByteArray): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

    /**
     * RFC 4122 big-endian: one byte per pair of hex digits from the canonical UUID,
     * matching C# WriteUuidBigEndian (Guid.ToString("N") then parse hex).
     */
    internal fun uuidBytes(canonical: String): ByteArray {
        val hex = canonicalUuid(canonical).replace("-", "")
        return ByteArray(16) { index ->
            hex.substring(index * 2, index * 2 + 2).toInt(16).toByte()
        }
    }

    private fun canonicalUuid(value: String): String {
        val lowered = value.lowercase()
        require(CANONICAL_UUID.matches(lowered)) { "UUID must be a canonical lowercase RFC 4122 string." }
        return lowered
    }

    private val CANONICAL_UUID = Regex(
        "^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$",
    )
}

File: windows/ClipSync.Core/Security/PairAuthProof.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace ClipSync.Core.Security;

/// <summary>
/// Pair-secret challenge-response proof from docs/protocol-v1.md section 3.
/// The shared reference vectors live in protocol/v1/fixtures/auth/vectors.json.
/// </summary>
public static class PairAuthProof
{
    public const int SecretLength = 32;
    public const int NonceLength = 32;
    public const int ProofLength = 32;

    private static readonly byte[] PrefixV1 = "ClipSync/v1/auth\n"u8.ToArray();
    private static readonly byte[] PrefixV2 = "ClipSync/v2/auth\n"u8.ToArray();

    public static byte[] Compute(
        ReadOnlySpan<byte> pairSecret,
        Guid challengeRequestId,
        ReadOnlySpan<byte> nonce,
        Guid challengerDeviceId,
        Guid responderDeviceId,
        long trustEpoch) =>
        Compute(pairSecret, challengeRequestId, nonce, challengerDeviceId, responderDeviceId, trustEpoch, protocolVersion: 1);

    public static byte[] Compute(
        ReadOnlySpan<byte> pairSecret,
        Guid challengeRequestId,
        ReadOnlySpan<byte> nonce,
        Guid challengerDeviceId,
        Guid responderDeviceId,
        long trustEpoch,
        int protocolVersion)
    {
        if (pairSecret.Length != SecretLength)
        {
            throw new ArgumentException("The pair secret must be exactly 32 bytes.", nameof(pairSecret));
        }

        if (nonce.Length != NonceLength)
        {
            throw new ArgumentException("The challenge nonce must be exactly 32 bytes.", nameof(nonce));
        }

        if (protocolVersion is not (1 or 2))
        {
            throw new ArgumentOutOfRangeException(nameof(protocolVersion), "Protocol version must be 1 or 2.");
        }

        var prefix = protocolVersion == 2 ? PrefixV2 : PrefixV1;
        var requestIdBytes = Encoding.UTF8.GetBytes(challengeRequestId.ToString("D"));
        var extra = protocolVersion == 2 ? 1 + 8 : 0;
        var message = new byte[prefix.Length + requestIdBytes.Length + 1 + NonceLength + 16 + 16 + 8 + extra];
        var offset = 0;

        prefix.CopyTo(message, offset);
        offset += prefix.Length;
        requestIdBytes.CopyTo(message, offset);
        offset += requestIdBytes.Length;
        message[offset] = 0x00;
        offset += 1;
        nonce.CopyTo(message.AsSpan(offset));
        offset += NonceLength;
        WriteUuidBigEndian(challengerDeviceId, message.AsSpan(offset, 16));
        offset += 16;
        WriteUuidBigEndian(responderDeviceId, message.AsSpan(offset, 16));
        offset += 16;
        BinaryPrimitives.WriteInt64BigEndian(message.AsSpan(offset, 8), trustEpoch);
        offset += 8;
        if (protocolVersion == 2)
        {
            message[offset] = 0x00;
            offset += 1;
            BinaryPrimitives.WriteInt64BigEndian(message.AsSpan(offset, 8), protocolVersion);
        }

        return HMACSHA256.HashData(pairSecret, message);
    }

    public static bool Verify(
        ReadOnlySpan<byte> pairSecret,
        Guid challengeRequestId,
        ReadOnlySpan<byte> nonce,
        Guid challengerDeviceId,
        Guid responderDeviceId,
        long trustEpoch,
        ReadOnlySpan<byte> proof)
    {
        if (proof.Length != ProofLength)
        {
            return false;
        }

        return Verify(pairSecret, challengeRequestId, nonce, challengerDeviceId, responderDeviceId, trustEpoch, proof, protocolVersion: 1);
    }

    public static bool Verify(
        ReadOnlySpan<byte> pairSecret,
        Guid challengeRequestId,
        ReadOnlySpan<byte> nonce,
        Guid challengerDeviceId,
        Guid responderDeviceId,
        long trustEpoch,
        ReadOnlySpan<byte> proof,
        int protocolVersion)
    {
        if (proof.Length != ProofLength)
        {
            return false;
        }

        var expected = Compute(
            pairSecret,
            challengeRequestId,
            nonce,
            challengerDeviceId,
            responderDeviceId,
            trustEpoch,
            protocolVersion);
        return CryptographicOperations.FixedTimeEquals(expected, proof);
    }

    /// <summary>RFC 4122 big-endian byte order, matching UUID_BYTES in the protocol document.</summary>
    private static void WriteUuidBigEndian(Guid value, Span<byte> destination)
    {
        var hex = value.ToString("N");
        for (var index = 0; index < 16; index++)
        {
            destination[index] = byte.Parse(hex.AsSpan(index * 2, 2), System.Globalization.NumberStyles.HexNumber, System.Globalization.CultureInfo.InvariantCulture);
        }
    }
}
